enclave/src/lib.rs

```diff
@@ -536,6 +539,69 @@ pub unsafe extern "C" fn run_db_thread() -> sgx_status_t {
     }
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn send_cid(
+    genesis_hash: *const u8,
+    genesis_hash_size: u32,
+    nonce: u32,
+    cid: *const u8,
+    cid_size: u32,
+    w_url: *const u8,
+    w_url_size: u32,
+    unchecked_extrinsic: *mut u8,
+    unchecked_extrinsic_size: u32,
+    new_size: *mut u32,
+) -> sgx_status_t {
+    // our certificate is unlinkable
+    let sign_type = sgx_quote_sign_type_t::SGX_UNLINKABLE_SIGNATURE;
+
+    info!("    [Enclave] Compose extrinsic");
+    let genesis_hash_slice = slice::from_raw_parts(genesis_hash, genesis_hash_size as usize);
+    //let mut nonce_slice     = slice::from_raw_parts(nonce, nonce_size as usize);
+    let url_slice = slice::from_raw_parts(w_url, w_url_size as usize);
+    let mut extrinsic_slice =
+        slice::from_raw_parts_mut(unchecked_extrinsic, unchecked_extrinsic_size as usize);
+    let signer = match ed25519::unseal_pair() {
+        Ok(pair) => pair,
+        Err(status) => return status,
+    };
+    info!("[Enclave] Restored ECC pubkey: {:?}", signer.public());
+
+    let genesis_hash = hash_from_slice(genesis_hash_slice);
+    debug!("decoded genesis_hash: {:?}", genesis_hash_slice);
+    debug!("worker url: {}", std::str::from_utf8(url_slice).unwrap());
+
+    let data: Vec<u8> = Vec::from(slice::from_raw_parts(cid, cid_size as usize));
+    let call: OpaqueCall = OpaqueCall(([OCEX_MODULE, OCEX_UPLOAD_CID], data.clone()).encode());
```

Comment:
Don't the call. The whole extrinsic will be encoded (see line 588) but not the call. Encoding the data (if not already encoded?) is enough.